<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Timothy Daley" />

<meta name="date" content="2020-11-29" />

<title>Fisher vs Stouffer</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/yeti.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/anchor-sections-1.0/anchor-sections.css" rel="stylesheet" />
<script src="site_libs/anchor-sections-1.0/anchor-sections.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 45px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 50px;
  margin-top: -50px;
}
.section h2 {
  padding-top: 50px;
  margin-top: -50px;
}
.section h3 {
  padding-top: 50px;
  margin-top: -50px;
}
.section h4 {
  padding-top: 50px;
  margin-top: -50px;
}
.section h5 {
  padding-top: 50px;
  margin-top: -50px;
}
.section h6 {
  padding-top: 50px;
  margin-top: -50px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">HOME</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="FisherVsStouffer.html">Fisher vs Stouffer</a>
</li>
<li>
  <a href="twoGroupsBackground.html">2 groups</a>
</li>
<li>
  <a href="empiricalFDR.html">empirical estimation of FDRs</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Fisher vs Stouffer</h1>
<h4 class="author">Timothy Daley</h4>
<h4 class="date">11/29/2020</h4>

</div>


<p>The most common way to coalesce <span class="math inline">\(p\)</span>-values is Fisher’s method (<a href="https://en.wikipedia.org/wiki/Fisher%27s_method" class="uri">https://en.wikipedia.org/wiki/Fisher%27s_method</a>). The basic idea is to log transform your <span class="math inline">\(p\)</span>-values, then <span class="math inline">\(-2\)</span> times the sum is chi-square distributed. Specifically, for <span class="math inline">\(p\)</span>-values <span class="math inline">\(p_{1}, \ldots, p_{n}\)</span>, Fisher’s method calculates the combined test statistic as <span class="math display">\[
t_{\text{Fisher}} = -2 \sum_{i = 1}^{n} \log p_{i}.
\]</span> Under the null that the <span class="math inline">\(p\)</span>-values are independent uniformly distributed, the test statistic <span class="math inline">\(t_{\text{Fisher}}\)</span> will be <span class="math inline">\(\chi^{2}_{2n}\)</span> distributed. The common wisdom is that Fisher’s method tend to values low p-values, instead of consistently low <span class="math inline">\(p\)</span>-values.</p>
<p>An alternative to Fisher’s method is Stouffer’s method. The idea is to transform the <span class="math inline">\(p\)</span>-values to <span class="math inline">\(z\)</span>-scores, then compute a combined <span class="math inline">\(z\)</span>-score by averaging the individual <span class="math inline">\(z\)</span>-scores. E.g. if <span class="math inline">\(\Phi(\cdot)\)</span> is the standard normal cdf, then the test statistic is <span class="math display">\[
t_{\text{Stouffer}} = \frac{1}{\sqrt{k}} \sum_{i = 1}^{n} \Phi^{-1} (p_{i}).
\]</span> Under the null this is distributed as a negative standard normal variable (negative because <span class="math inline">\(\Phi^{-1}\)</span> of small <span class="math inline">\(p\)</span>-values will be less than zero). The prevailing wisdom is that this method values consistently small <span class="math inline">\(p\)</span>-values. To ensure that effects have the same sign, we can define <span class="math display">\[
t_{\text{Stouffer}} = \frac{1}{\sqrt{k}} \sum_{i = 1}^{n} \text{sign}(x_{i}) \Phi^{-1} (p_{i})
\]</span> as the test statistic so that effect sizes (<span class="math inline">\(x_{i}\)</span>) of the same sign are valued. This has a null distribution of standard normal.</p>
<p>I first learned about Stouffer’s method through a <a href="https://projecteuclid.org/download/pdfview_1/euclid.aos/1256303530">paper by Art Owen</a> which discusses different meta-analysis techniques. In particular, he says “In the Fisher test, if the first <span class="math inline">\(m−1\)</span> <span class="math inline">\(p\)</span>-values already yield a test statistic exceeding the <span class="math inline">\(\chi^2_{2m}\)</span> significance threshold, then the <span class="math inline">\(m\)</span>th test statistic cannot undo it. The Stouffer test is different. Any large but finite value of <span class="math inline">\(\sum_{j = 1}^{m-1} \Phi^{-1}(p_{j})\)</span> can be canceled by an opposing value of <span class="math inline">\(\Phi^{-1}(p_{m})\)</span>.” Thus indicating that Stouffer’s method will tend to pick up more small but consistent effects, and will be more robust in the presence of rare outliers. On the other hand, because Fisher’s method can be significant due to just one very small <span class="math inline">\(p\)</span>-value then it is likely not as robust to outliers. To test this let’s compare the methods in two situations: small-tailed null and long-tailed null, where the latter will simulate the presence of outlier effects. I expect that Fisher’s method will have higher false positive in the presence of long tails.</p>
<div id="simulation" class="section level1">
<h1>Simulation</h1>
<p>The simulation I’ll run is the standard 2-groups model (e.g. see <a href="https://projecteuclid.org/download/pdfview_1/euclid.ss/1215441276">Efron 2008</a>). For aggregation purposes, we’ll have 5 effect sizes (I’m leaving effect size undefined on purpose) to merge. In the motivating application, the individual effect sizes and <span class="math inline">\(p\)</span>-values arise from differential expression analysis of guide RNAs, with aggregation done at the gene level (as guide RNAs target specific genes). Consistent effects are not guaranteed because of variable guide efficiency (e.g. see <a href="https://link.springer.com/article/10.1186/s13059-018-1538-6">my previous paper on this topic</a>).</p>
<pre class="r"><code>library(reticulate)
# you need to edit RETICULATE_PYTHON in .Renviron
# for some reason I need to make a venv for each project
Sys.setenv(RETICULATE_PYTHON = &quot;/Users/tim.daley/blog/FisherVsStouffer/venv/bin/python3&quot;)
use_python(&#39;/Users/tim.daley/blog/FisherVsStouffer/venv/bin/python3&#39;)
#reticulate::py_config()</code></pre>
<pre class="python"><code>import numpy as np
# set seed
np.random.seed(12345)
n_genes = 20000
# 2% of genes are non-null
p = 0.05
gene_labels = np.random.binomial(n = 1, p = p, size = (n_genes, ))
print(&quot;number of genes: &quot;, n_genes)</code></pre>
<pre><code>## number of genes:  20000</code></pre>
<pre class="python"><code>print(&quot;number of non-null genes: &quot;, sum(gene_labels))</code></pre>
<pre><code>## number of non-null genes:  1029</code></pre>
<div id="small-tail" class="section level2">
<h2>Small tail</h2>
<p>For the small-tailed case, let’s assume that null gene-effects are equal to <span class="math inline">\(0\)</span> and the non-null gene effects are distributed <span class="math inline">\(\mathcal{N}(3, 1)\)</span>.</p>
<pre class="python"><code>import seaborn
gene_effects = gene_labels*np.random.normal(loc = 3, scale = 1, size = (n_genes, ))
seaborn.histplot(gene_effects, color = &#39;black&#39;)</code></pre>
<p><img src="FisherVsStouffer_files/figure-html/unnamed-chunk-2-1.png" width="672" /></p>
<p>You can see the long tail on the right is the gene effects. Now let’s assume 5 observations (guide RNAs) per gene. These are gonna be normally distributed with mean equal to gene effect size, and standard deviation 1.</p>
<pre class="python"><code>import itertools
import pandas as pd

lst = range(n_genes)
guide2gene_map = list(itertools.chain.from_iterable(itertools.repeat(x, 5) for x in lst))
n_guides = 5*n_genes
assert(len(guide2gene_map) == n_guides)

guide_effects = pd.DataFrame({&#39;guide2gene&#39;: guide2gene_map})
guide_effects[&#39;guide_effect&#39;] = guide_effects[&#39;guide2gene&#39;].map(lambda i: np.random.normal(loc = gene_effects[i], scale = 1))
guide_effects[&#39;label&#39;] = guide_effects[&#39;guide2gene&#39;].map(lambda i: gene_labels[i])
seaborn.histplot(x = &#39;guide_effect&#39;, hue = &#39;label&#39;, data = guide_effects).set_title(&#39;guide effect distribution&#39;)</code></pre>
<p><img src="FisherVsStouffer_files/figure-html/unnamed-chunk-3-1.png" width="672" /></p>
<p>Now let’s look at combining <span class="math inline">\(p\)</span>-values with Fisher vs Stouffer.</p>
<pre class="python"><code>from scipy.stats import norm, chi2
from math import sqrt

def stouffer_pval(z):
  stouffer_zval = sum(z)/sqrt(len(z))
  stouffer_pval = norm.cdf(-stouffer_zval)
  return stouffer_pval

def fisher_pval(z):
  # 1-sided p-val
  pvals = norm.cdf(-z)
  fisher_val = -2*sum(np.log(pvals))
  fisher_pval = 1.0 - chi2.cdf(fisher_val, df = 2*len(z))
  return fisher_pval
  
gene_pvals = pd.DataFrame({&#39;gene&#39;: range(n_genes),
                           &#39;gene_label&#39;: gene_labels})
                           
gene_pvals[&#39;fisher&#39;] = gene_pvals[&#39;gene&#39;].map(lambda i: fisher_pval(guide_effects[guide_effects[&#39;guide2gene&#39;] == gene_pvals[&#39;gene&#39;][i]][&#39;guide_effect&#39;]))
gene_pvals[&#39;stouffer&#39;] = gene_pvals[&#39;gene&#39;].map(lambda i: stouffer_pval(guide_effects[guide_effects[&#39;guide2gene&#39;] == gene_pvals[&#39;gene&#39;][i]][&#39;guide_effect&#39;]))

gene_pvals.head()</code></pre>
<pre><code>##    gene  gene_label    fisher  stouffer
## 0     0           0  0.620273  0.508737
## 1     1           0  0.873517  0.868306
## 2     2           0  0.509792  0.379961
## 3     3           0  0.372420  0.446334
## 4     4           0  0.398732  0.542383</code></pre>
<pre class="python"><code>import matplotlib.pyplot as plt
seaborn.histplot(gene_pvals[&#39;fisher&#39;], bins = 30, color = &#39;black&#39;, label = &#39;Fisher p-vals&#39;)
plt.show()</code></pre>
<p><img src="FisherVsStouffer_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
<pre class="python"><code>seaborn.histplot(gene_pvals[&#39;stouffer&#39;], bins = 30, color = &#39;black&#39;, label = &#39;Stouffer p-vals&#39;)
plt.show()</code></pre>
<p><img src="FisherVsStouffer_files/figure-html/unnamed-chunk-5-2.png" width="672" /></p>
<p>Now let’s compute Benjamini-Hochberg FDRs and compare them to the empirical FDR.</p>
<pre class="python"><code>from statsmodels.stats.multitest import fdrcorrection

_, gene_pvals[&#39;fisher_fdr&#39;] = fdrcorrection(gene_pvals[&#39;fisher&#39;], method = &#39;indep&#39;)

_, gene_pvals[&#39;stouffer_fdr&#39;] = fdrcorrection(gene_pvals[&#39;stouffer&#39;], method = &#39;indep&#39;)

gene_pvals.head()</code></pre>
<pre><code>##    gene  gene_label    fisher  stouffer  fisher_fdr  stouffer_fdr
## 0     0           0  0.620273  0.508737    0.970138      0.953773
## 1     1           0  0.873517  0.868306    0.990323      0.991374
## 2     2           0  0.509792  0.379961    0.956906      0.921007
## 3     3           0  0.372420  0.446334    0.920225      0.940338
## 4     4           0  0.398732  0.542383    0.926149      0.956265</code></pre>
<p>Now let’s compute the empirical FDRs.</p>
<pre class="python"><code>def compute_empirical_fdr(labels, fdrs, fdr_thresh):
  mask = (fdrs &lt; fdr_thresh)
  R = np.sum(mask)
  if R &gt; 0:
    return np.sum(1 - labels[mask])/float(R)
  else:
    return 0

empirical_fdr = pd.DataFrame({&#39;fdr_thresh&#39;: np.linspace(0, 1, num = 1000)})  
empirical_fdr[&#39;fisher&#39;] = empirical_fdr[&#39;fdr_thresh&#39;].map(lambda x: compute_empirical_fdr(gene_pvals[&#39;gene_label&#39;], gene_pvals[&#39;fisher_fdr&#39;], x))

empirical_fdr[&#39;stouffer&#39;] = empirical_fdr[&#39;fdr_thresh&#39;].map(lambda x: compute_empirical_fdr(gene_pvals[&#39;gene_label&#39;], gene_pvals[&#39;stouffer_fdr&#39;], x))

empirical_fdr.head()</code></pre>
<pre><code>##    fdr_thresh    fisher  stouffer
## 0    0.000000  0.000000  0.000000
## 1    0.001001  0.001147  0.001121
## 2    0.002002  0.002227  0.003293
## 3    0.003003  0.004410  0.004362
## 4    0.004004  0.007650  0.007559</code></pre>
<pre class="python"><code>empirical_fdr.tail()</code></pre>
<pre><code>##      fdr_thresh    fisher  stouffer
## 995    0.995996  0.945746  0.945044
## 996    0.996997  0.946220  0.945427
## 997    0.997998  0.948362  0.947968
## 998    0.998999  0.948434  0.948398
## 999    1.000000  0.948550  0.948550</code></pre>
<pre class="python"><code>df = pd.DataFrame({&#39;estimated_fdr&#39;: empirical_fdr[&#39;fdr_thresh&#39;].tolist() + empirical_fdr[&#39;fdr_thresh&#39;].tolist(), 
                   &#39;empirical_fdr&#39;: empirical_fdr[&#39;fisher&#39;].tolist() + empirical_fdr[&#39;stouffer&#39;].tolist(),
                   &#39;condition&#39;: [&#39;fisher&#39;]*empirical_fdr.shape[0] + [&#39;stouffer&#39;]*empirical_fdr.shape[0]})

plt.plot([0, 1], [0, 1], &#39;--&#39;, color = &#39;black&#39;)</code></pre>
<pre><code>## [&lt;matplotlib.lines.Line2D object at 0x7ffd20a440d0&gt;]</code></pre>
<pre class="python"><code>seaborn.lineplot(x = &#39;estimated_fdr&#39;, y = &#39;empirical_fdr&#39;, hue = &#39;condition&#39;, data = df)</code></pre>
<pre><code>## &lt;AxesSubplot:xlabel=&#39;estimated_fdr&#39;, ylabel=&#39;empirical_fdr&#39;&gt;</code></pre>
<pre class="python"><code>plt.show()</code></pre>
<p><img src="FisherVsStouffer_files/figure-html/unnamed-chunk-9-1.png" width="672" /></p>
</div>
<div id="long-tail" class="section level2">
<h2>Long tail</h2>
<p>Now what happens if the null is a long-tailed distribution? We’ll use a t-distribution with a finite variance as the long tailed, and we’ll normalize it so that the standard deviation is equal to 1.</p>
<pre class="python"><code>import pandas as pd
from math import sqrt
long_tailed_guide_effects = pd.DataFrame({&#39;guide2gene&#39;: guide2gene_map})
deg_fred = 5
sd = sqrt(deg_fred/float(deg_fred - 2))
# divide by sd to normalize
long_tailed_guide_effects[&#39;guide_effect&#39;] = long_tailed_guide_effects[&#39;guide2gene&#39;].map(lambda i: gene_effects[i] + np.random.standard_t(df = deg_fred)/sd)
long_tailed_guide_effects[&#39;label&#39;] = long_tailed_guide_effects[&#39;guide2gene&#39;].map(lambda i: gene_labels[i])
seaborn.histplot(x = &#39;guide_effect&#39;, hue = &#39;label&#39;, data = long_tailed_guide_effects)</code></pre>
<pre><code>## &lt;AxesSubplot:xlabel=&#39;guide_effect&#39;, ylabel=&#39;Count&#39;&gt;</code></pre>
<pre class="python"><code>plt.show()</code></pre>
<p><img src="FisherVsStouffer_files/figure-html/unnamed-chunk-10-1.png" width="672" /></p>
<pre class="python"><code>long_tailed_gene_pvals = pd.DataFrame({&#39;gene&#39;: range(n_genes),
                           &#39;gene_label&#39;: gene_labels})
                           
long_tailed_gene_pvals[&#39;fisher&#39;] = long_tailed_gene_pvals[&#39;gene&#39;].map(lambda i: fisher_pval(long_tailed_guide_effects[long_tailed_guide_effects[&#39;guide2gene&#39;] == long_tailed_gene_pvals[&#39;gene&#39;][i]][&#39;guide_effect&#39;]))
long_tailed_gene_pvals[&#39;stouffer&#39;] = long_tailed_gene_pvals[&#39;gene&#39;].map(lambda i: stouffer_pval(long_tailed_guide_effects[long_tailed_guide_effects[&#39;guide2gene&#39;] == long_tailed_gene_pvals[&#39;gene&#39;][i]][&#39;guide_effect&#39;]))
_, long_tailed_gene_pvals[&#39;fisher_fdr&#39;] = fdrcorrection(long_tailed_gene_pvals[&#39;fisher&#39;], method = &#39;indep&#39;)

_, long_tailed_gene_pvals[&#39;stouffer_fdr&#39;] = fdrcorrection(long_tailed_gene_pvals[&#39;stouffer&#39;], method = &#39;indep&#39;)

long_tailed_empirical_fdr = pd.DataFrame({&#39;fdr_thresh&#39;: np.linspace(0, 1, num = 1000)})  
long_tailed_empirical_fdr[&#39;fisher&#39;] = long_tailed_empirical_fdr[&#39;fdr_thresh&#39;].map(lambda x: compute_empirical_fdr(long_tailed_gene_pvals[&#39;gene_label&#39;], long_tailed_gene_pvals[&#39;fisher_fdr&#39;], x))
long_tailed_empirical_fdr[&#39;stouffer&#39;] = long_tailed_empirical_fdr[&#39;fdr_thresh&#39;].map(lambda x: compute_empirical_fdr(long_tailed_gene_pvals[&#39;gene_label&#39;], long_tailed_gene_pvals[&#39;stouffer_fdr&#39;], x))

df = pd.DataFrame({&#39;estimated_fdr&#39;: long_tailed_empirical_fdr[&#39;fdr_thresh&#39;].tolist() + long_tailed_empirical_fdr[&#39;fdr_thresh&#39;].tolist(), 
                   &#39;empirical_fdr&#39;: long_tailed_empirical_fdr[&#39;fisher&#39;].tolist() + long_tailed_empirical_fdr[&#39;stouffer&#39;].tolist(),
                   &#39;condition&#39;: [&#39;fisher&#39;]*long_tailed_empirical_fdr.shape[0] + [&#39;stouffer&#39;]*long_tailed_empirical_fdr.shape[0]})</code></pre>
<pre class="python"><code>plt.plot([0, 1], [0, 1], &#39;--&#39;, color = &#39;black&#39;)</code></pre>
<pre><code>## [&lt;matplotlib.lines.Line2D object at 0x7ffd1083e070&gt;]</code></pre>
<pre class="python"><code>seaborn.lineplot(x = &#39;estimated_fdr&#39;, y = &#39;empirical_fdr&#39;, hue = &#39;condition&#39;, data = df)</code></pre>
<pre><code>## &lt;AxesSubplot:xlabel=&#39;estimated_fdr&#39;, ylabel=&#39;empirical_fdr&#39;&gt;</code></pre>
<pre class="python"><code>plt.show()</code></pre>
<p><img src="FisherVsStouffer_files/figure-html/unnamed-chunk-12-1.png" width="672" /></p>
<p>We see here that Fisher’s method has a higher false positive rate, particularly in the lower region where we usually set the false discovery rate. This inidicates that Fisher’s method is much more susceptible to false positives in the presence of long tails. If we correctly specify the null model, then this will be less of an issue, but we saw above that both methods perform similarly when the model is correctly specified. Therefore, this simple experiment indicates that there can be significant benefits to using Stouffer’s method in place of Fisher’s method, particularly when long-tailed or off-target effects can be present in the null distribution and are difficult to accurately model.</p>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
