<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Timothy Daley" />


<title>Bias of the Gini coefficient from Poisson sampling</title>

<script src="site_libs/header-attrs-2.13/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/yeti.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">HOME</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="FisherVsStouffer.html">Fisher vs Stouffer</a>
</li>
<li>
  <a href="twoGroupsBackground.html">2 groups</a>
</li>
<li>
  <a href="empiricalFDR.html">empirical estimation of FDRs</a>
</li>
<li>
  <a href="CRISPR_qc/logNormalPoissonGoodTuring.html">Good-Turing as an alternative to adding 1</a>
</li>
<li>
  <a href="CRISPR_qc/CorrelationBreakdown.html">signal to noise ratio in CRISPR pooled screens</a>
</li>
<li>
  <a href="crispr_tko/tree_model/tree_model.html">What's a test set?</a>
</li>
<li>
  <a href="IntroToCRISPRiaScreenAnalysis/Intro2CRISPRScreenAnalysis.html">An intro to the analysis of CRISPR pooled screens</a>
</li>
<li>
  <a href="EpicBlog/EngineeringInterviewBlogPost.html">Bioinformatics Engineer Exam</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Bias of the Gini coefficient from Poisson
sampling</h1>
<h4 class="author">Timothy Daley</h4>
<h4 class="date">7/15/2022</h4>

</div>


<p>The Gini coefficient is commonly used to measure the inequality of a
population, particularly in economic contexts. For example, the Gini
coefficient is commonly used to measure and compare wealth or income
inequality across countries or areas (see <a
href="https://en.wikipedia.org/wiki/Gini_coefficient">Wikipedia</a>).
The Gini coefficent is defined as <span class="math display">\[
\frac{ \sum_i \sum_j | x_i - x_j |}{2 n \sum_j x_j}.
\]</span> The Gini coefficient ranges from 0, when all <span
class="math inline">\(x_i = x_j\)</span> and the population is
completely uniform, to 1, when some <span class="math inline">\(x_k &gt;
0\)</span> and all other <span class="math inline">\(x_{i} = 0\)</span>,
i.e. one person has all the wealth.</p>
<p>The Gini coefficient is used in genomic contexts to compare
inequality of a sampled population, e.g. <a
href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-1010-4"
class="uri">https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-1010-4</a>
or <a
href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-015-0843-6"
class="uri">https://genomebiology.biomedcentral.com/articles/10.1186/s13059-015-0843-6</a>.
The latter paper is where our motivation comes from. The Gini
coefficient is used to assess the quality or selection of sequenced
CRISPR screening libraries. Where the usual Gini coefficients assess the
distribution of sampled quantities, e.g. sampling random people in a
population and obtaining the income for each one, our motivating
application involves Poisson or multinomial sampling from a population
to assess the frequency. With the latter, there is a sampling-based bias
that can affect our estimates.</p>
<p>To see this point, consider the extreme situation where the
population being sampled is completely uniform. The true Gini
coefficient is equal to zero, however because the observed sample is not
completely uniform we get a stricly positive estimated Gini coefficient.
Therefore we have some sampling-based bias. How much exactly? We shall
see in this investigation.</p>
<div id="simulation" class="section level1">
<h1>Simulation</h1>
<p>We’ll do some simulations to determine the rate of bias in the
Poisson case.</p>
<pre class="r"><code>gini &lt;- function(x){
  if(length(x) == 0){
    return(0)
  }
  if(sum(x) == 0){
    return(1)
  }
  if(!all(x &gt;= 0)){
    return(1)
  }
  x.sorted = sort(x)
  l = length(x)
  numer = sum((l - 1:l + 1)*x.sorted)
  denom = sum(x.sorted)
  return((l + 1 - 2*numer/denom)/l)
}
n_rep = 100
n_sgrna = 200000
sampling_depth = n_sgrna*c(1/16, 1/8, 1/4, 1/2, 1, 2, 4, 8, 16, 32) 
sampled_gini = data.frame(sampling_depth = rep(sampling_depth, each = n_rep),
                          gini = rep(0, times = length(sampling_depth)*n_rep))
for(i in 1:dim(sampled_gini)[1]){
 lam = sampled_gini$sampling_depth[i]
 x = rpois(n_sgrna, lambda = lam)
 sampled_gini$gini[i] = gini(x)
}

library(dplyr)
gini_summary = sampled_gini %&gt;% group_by(sampling_depth) %&gt;% summarize(gini = mean(gini))
library(ggplot2)
ggplot(gini_summary, aes(x = sampling_depth, y = gini)) + geom_line() + theme_bw()</code></pre>
<p><img src="gini_files/figure-html/unnamed-chunk-1-1.png" width="672" /></p>
<pre class="r"><code>ggplot(gini_summary, aes(x = sampling_depth, y = gini)) + geom_line() + theme_bw() + scale_y_log10() + scale_x_log10()</code></pre>
<p><img src="gini_files/figure-html/unnamed-chunk-1-2.png" width="672" /></p>
<p>We would naturally think that the convergence rate is some <span
class="math inline">\(O(n^p)\)</span>, e.g. <span
class="math inline">\(O(n^{1/2})\)</span>, as we see from the above
<span class="math inline">\(\log - \log\)</span> plot.</p>
<pre class="r"><code>summary(lm(log(sampling_depth) ~ log(gini), data = gini_summary))</code></pre>
<pre><code>## 
## Call:
## lm(formula = log(sampling_depth) ~ log(gini), data = gini_summary)
## 
## Residuals:
##        Min         1Q     Median         3Q        Max 
## -4.340e-04 -1.449e-04 -2.100e-07  1.369e-04  5.385e-04 
## 
## Coefficients:
##               Estimate Std. Error t value Pr(&gt;|t|)    
## (Intercept) -1.1448271  0.0007031   -1628   &lt;2e-16 ***
## log(gini)   -2.0000006  0.0001016  -19686   &lt;2e-16 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Residual standard error: 0.0003198 on 8 degrees of freedom
## Multiple R-squared:      1,  Adjusted R-squared:      1 
## F-statistic: 3.875e+08 on 1 and 8 DF,  p-value: &lt; 2.2e-16</code></pre>
<p>We see that the convergence rate of the error is <span
class="math inline">\(O(n^{-2})\)</span> in the uniform Poisson case.
But what if the we’re not in the perfectly uniform case, as we’re
commonly not.</p>
<p>We’ll get to that next time.</p>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
